### PR TITLE
Fix Lots of Warnings

### PR DESCRIPTION
**Pull request by wesbland**
## Pull Request Description

This is a grab bag of warning fixes for things not currently covered by the warning checker job. Most of this is related to unused labels that show up when you compile with error checking off (i.e., all of the `MPIR_ERR_CHECK` and `MPIR_Assert` macros disappear), but there are a few others.

Most of these can be reproduced by using `--enable-strict` and `--disable-error-checking`. A few others come from using ICC/ICX instead of GCC.

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by wesbland** [comment]
Ok. I"m done messing with this. It should be ready for review now @travisk1154.

**Review by hzhou — COMMENTED** [review]
All the changes look okay to me. But I have to admit I didn't spend much time at all looking into them. I think you can go ahead group the commits by type of changes and type of files, into maybe under 10 or so commits. Meanwhile, I'll ask Ken to also look at them.

**Comment by wesbland** [comment]
I squashed a few of these together and grouped the rest together by location in code. Only three are longer than a few lines:

* 4e7f9271c75d - mpi/coll: Suppress unused variable warning
* aeae7eb93e12 - mpl: Cleanup return labels
* fb0fc4628faf - mpl: Fix function definition

**Comment by wesbland** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by travisk1154** [comment]
Yea my comment was going to be similar to Hui's. If not too much, would you be able to add the flags in the PR description that you used to cause the warnings? Otherwise LGTM

**Comment by wesbland** [comment]
There's a bunch of configs that I used to generate these, but the main thing I did was `--enable-strict` and `--disable-error-checking`. The covers pretty much all of the "unused something" warnings. Some of them came from using ICC/ICX instead of GCC.

At some point, I did give up on getting the labels right because it just gets too complex. That's when I noticed that at some point in the past, we at least said we were going to turn off `-Wunused-labels` in the strict build.

**Comment by hzhou** [comment]
>  That's when I noticed that at some point in the past, we at least said we were going to turn off `-Wunused-labels` in the strict build.

We did. But `icc` doesn't have that option. And pgi compilers and suncc also had trouble suppressing it. So at one point I decided to just fix the labels. I wasn't thinking about the `--disable-error-checking`. I guess that makes the issue unfixable, huh?

**Comment by wesbland** [comment]
> We did. But `icc` doesn't have that option. And pgi compilers and suncc also had trouble suppressing it. So at one point I decided to just fix the labels. I wasn't thinking about the `--disable-error-checking`. I guess that makes the issue unfixable, huh?

It sounds like that should be resolved by having different sets of strict flags for different compilers rather than reverting to the least common denominator. I have access to ICC (obviously), but not PGI or Sun CC. Is that something you can help with?

**Review by travisk1154 — APPROVED** [review]
Thanks for updating the PR description. LGTM.

**Comment by wesbland** [comment]
I potentially have a fix for the different compilers for strict builds coming. Give me a few minutes to test it out.

**Comment by wesbland** [comment]
@hzhou Please take a look at the last patch.

File: confdb/aclocal_cc.m4 [hunk 1]
```diff
@@ -403,6 +403,56 @@ else
...
+    PAC_RUNLOG([$CC $arg </dev/null >conftest.txt 2>&1])
+    # Ignore the return code, because some compilers set the
+    # return code to zero on invalid arguments and some to
+    # non-zero on success (with no files to compile)
+    if test -f conftest.txt ; then
+        if grep 'Portland Group' conftest.txt >/dev/null 2>&1 ; then
+            pac_cv_cc_vendor=pgi
+        elif grep 'Sun Workshop' conftest.txt >/dev/null 2>&1 ; then
+            pac_cv_cc_vendor=sun
+	elif grep 'Sun C' conftest.txt >/dev/null 2>&1 ; then
```
Comment (hzhou) [inline confdb/aclocal_cc.m4:429]: What happened to the indentation? Does your editor fail you? 😄

File: confdb/aclocal_cc.m4 [hunk 2]
```diff
@@ -403,6 +403,56 @@ else
...
+if test "$pac_cv_cc_vendor" = "unknown" ; then
+    # Try to use the compiler name
+    if test "$CC" = "icc" -o "$CC" = "icx" ; then
+        pac_cv_cc_vendor=intel
+    elif test "$CC" = "pgcc" ; then
+        pac_cv_cc_vendor=pgi
+    elif test "$CC" = "xlcc" ; then
+        pac_cv_cc_vendor=ibm
+    fi
+fi
```
Comment (hzhou) [inline confdb/aclocal_cc.m4:450]: Missing `clang`?

File: confdb/aclocal_cc.m4 [hunk 3]
```diff
@@ -508,6 +558,16 @@ if test "$enable_strict_done" != "yes" ; then
...
+    case "$pac_cv_cc_vendor" in
+        gnu)
+            pac_common_strict_flags="${pac_common_strict_flags}
+                -Wno-unused-label
+            "
+            ;;
+        *)
+            ;;
+    esac
+
```
Comment (hzhou) [inline confdb/aclocal_cc.m4:570]: Yeah, I think this solution is more straightforward. 👍

File: confdb/aclocal_cc.m4 (hunk 1 above)
Comment (wesbland) [reply confdb/aclocal_cc.m4:429]: Copy/pasting from the Fortran version with a mix of tabs and spaces failed me...

File: confdb/aclocal_cc.m4 (hunk 2 above)
Comment (wesbland) [reply confdb/aclocal_cc.m4:450]: I don't think that's really a problem. It'll stay as "unknown", which means it won't use the GCC-isms. I'll add something anyway...

**Comment by wesbland** [comment]
I've addressed the points above.

File: src/mpi/coll/algorithms/recexchalgo/recexchalgo.c [hunk 4]
```diff
@@ -272,7 +272,8 @@ int MPII_Recexchalgo_reverse_digits_step2(int rank, int comm_size, int k)
     int i, T, rem, power, step2rank, step2_reverse_rank = 0;
     int pofk = 1, log_pofk = 0;
     int *digit, *digit_reverse;
-    int mpi_errno = MPI_SUCCESS;
+    int mpi_errno ATTRIBUTE((unused));
+    mpi_errno = MPI_SUCCESS;
```
Comment (raffenet) [inline src/mpi/coll/algorithms/recexchalgo/recexchalgo.c:276]: nitpick: none of the other instances in this commit split the assignment onto a second line

File: src/mpi/coll/transports/gentran/tsp_gentran.c [hunk 5]
```diff
@@ -504,7 +504,7 @@ int MPIR_TSP_sched_sink(MPIR_TSP_sched_t s, int *vtx_id)
 int MPIR_TSP_sched_fence(MPIR_TSP_sched_t s)
 {
     MPII_Genutil_sched_t *sched = s;
-    int mpi_errno = MPI_SUCCESS;
+    int mpi_errno ATTRIBUTE((unused)) = MPI_SUCCESS;
```
Comment (raffenet) [inline src/mpi/coll/transports/gentran/tsp_gentran.c:507]: looks like this was leaked from another commit

File: src/mpid/ch4/src/ch4r_init.c [hunk 6]
```diff
@@ -17,11 +17,8 @@ int MPIDIG_init_comm(MPIR_Comm * comm)
     if (MPIR_CONTEXT_READ_FIELD(DYNAMIC_PROC, comm->recvcontext_id))
         goto fn_exit;
 
-    subcomm_type = MPIR_CONTEXT_READ_FIELD(SUBCOMM, comm->recvcontext_id);
-    is_localcomm = MPIR_CONTEXT_READ_FIELD(IS_LOCALCOMM, comm->recvcontext_id);
-
-    MPIR_Assert(subcomm_type <= 3);
-    MPIR_Assert(is_localcomm <= 1);
+    MPIR_Assert(MPIR_CONTEXT_READ_FIELD(SUBCOMM, comm->recvcontext_id) <= 3);
+    MPIR_Assert(MPIR_CONTEXT_READ_FIELD(IS_LOCALCOMM, comm->recvcontext_id) <= 1);
```
Comment (raffenet) [inline src/mpid/ch4/src/ch4r_init.c:21]: I think these asserts can just be removed. There used to be a context_id->comm lookup in this places that was removed.

File: src/mpid/ch4/src/ch4r_init.c [hunk 7]
```diff
@@ -31,16 +28,14 @@ int MPIDIG_init_comm(MPIR_Comm * comm)
...
 
     if (MPIR_CONTEXT_READ_FIELD(DYNAMIC_PROC, comm->recvcontext_id))
         goto fn_exit;
-    subcomm_type = MPIR_CONTEXT_READ_FIELD(SUBCOMM, comm->recvcontext_id);
-    is_localcomm = MPIR_CONTEXT_READ_FIELD(IS_LOCALCOMM, comm->recvcontext_id);
 
-    MPIR_Assert(subcomm_type <= 3);
-    MPIR_Assert(is_localcomm <= 1);
+    MPIR_Assert(MPIR_CONTEXT_READ_FIELD(SUBCOMM, comm->recvcontext_id) <= 3);
+    MPIR_Assert(MPIR_CONTEXT_READ_FIELD(IS_LOCALCOMM, comm->recvcontext_id) <= 1);
```
Comment (raffenet) [inline src/mpid/ch4/src/ch4r_init.c:38]: these can be removed

**Review by raffenet — COMMENTED** [review]
Made comments as I went. Just minor stuff.

File: src/mpi/coll/algorithms/recexchalgo/recexchalgo.c (hunk 4 above)
Comment (wesbland) [reply src/mpi/coll/algorithms/recexchalgo/recexchalgo.c:276]: Fixed. I learned that I could do that somewhere along the way and didn't backport my simplification.

File: src/mpi/coll/transports/gentran/tsp_gentran.c (hunk 5 above)
Comment (wesbland) [reply src/mpi/coll/transports/gentran/tsp_gentran.c:507]: Yep. I've moved it back.

File: src/mpid/ch4/src/ch4r_init.c (hunk 6 above)
Comment (wesbland) [reply src/mpid/ch4/src/ch4r_init.c:21]: Gone.

**Comment by wesbland** [comment]
@raffenet I took care of your comments and added 4 more commits that came up with further internal builds. In particular, the last commit is something you should take a look at because it basically turns off warning checking for submodules.

I think it's unfair to punish MPICH for the submodules warnings. They should have their own warning checkers if we care and in some cases, since we don't control them, it's not reasonable to expect them to be clean.

**Comment by wesbland** [comment]
I moved all of the `aclocal_cc.m4` commits to the beginning and squashed them where it made sense. I also differentiated between the Intel class compiler and the Intel oneAPI compiler. The later includes `-Wno-unused-labels` while the former does not. Take another quick look at that one before giving it the final 👍 please.

**Review by hzhou — APPROVED** [review]
LGTM

**Review by raffenet — APPROVED** [review]

**Pull request merged**